ccb/upstream_project.py

```diff
@@ -256,4 +256,54 @@ def _get_owner_repo(cls, recipe):
         raise _Unsupported()
 
 
-_CLASSES = [GithubProject]
+class GnomeProject(UpstreamProject):
+    SOURCE_URL_RE = re.compile(r"https?://((download.gnome.org)|(ftp.gnome.org/pub/gnome))/sources/([^/]+)/")
+
+    def __init__(self, recipe):
+        domain, _, _, project = self._get_project(recipe)
+        super().__init__(recipe)
+        self.domain = domain
+        self.project = project
+        self.__versions = None
+
+    @classmethod
+    def _get_project(cls, recipe):
+        try:
+            url = recipe.source()["url"]
+            match = cls.SOURCE_URL_RE.match(url)
+            if match:
+                return match.groups()
+        except Exception as exc:
+            logger.debug(
+                "%s: not supported as Gnome project because of the following exception: %s",
+                recipe.name,
+                repr(exc),
+            )
+        raise _Unsupported()
+
+    async def versions(self):
+        if self.__versions is None:
+            try:
+                async with aiohttp.ClientSession(raise_for_status=True) as client:
+                    async with client.get(f"https://{self.domain}/sources/{self.project}/cache.json") as resp:
+                        d = await resp.json()
+                        self.__versions = [Version(v) for v in d[2][self.project]]
+            except Exception as exc:
+                logger.info("%s: error parsing repository: %s", self.recipe.name, exc)
+                logger.debug(traceback.format_exc())
+                self.__versions = list()
+        return self.__versions
+
+    async def most_recent_version(self):
```

Comment:
Why do you need to define this function ? Isn't the same as in the base class ?

Reply:
The base class upstramproject does not have this implementation, but maybe it should be moved from gitproject class to upstreamproject?

Reply:
Oh right. It could indeed be in UpstreamProject, it's a good default implementation.
By the way I noticed the type hint for the return value of `versions()` is wrong in `UpstreamProject`